Add panel view and bulk commit operations

LeuciRemi opened the pull request:
My goal was to add git-graph in the panel section to keep an eye on commits. And to add some practical jetbrains functions like Undo and squash with multi select on commits.

## Global Changes
- Added support for viewing Git Graph in a dedicated panel or editor tab
<img width="1840" alt="image" src="https://github.com/user-attachments/assets/426bed6d-12fa-4d18-8b12-69d34bbf4f4b" />
<img width="1840" alt="image" src="https://github.com/user-attachments/assets/68a9b848-26ce-4663-af25-b3b9bf9b2d0b" />

---
- Added multiple commit selection to enable bulk operations
  - Added `Squash Selected Commits` action when multiple commits are selected
  - Added `Drop Selected Commits` action when multiple commits are selected
![image](https://github.com/user-attachments/assets/b2b73cac-1a22-4066-a66c-64366a7809d7)

## Commit menu features
- Added `Edit Message` action to commit context menu for modifying commit messages
- Added `Undo Last Commit` action to last commit context menu

## Branch menu features
- Added `Create Branch` action to branch context menu (same as `Create Branch` on a commit context menu)
- Added `Pull Branch` action to non-checked branch context menu

hansu commented:
Thanks for that big packet of features! Good job :+1: 

But I get a lot of ESLint warnings for your newly created files. I guess you are using the default git setting with `core.autocrlf = false`?

```bash
  ...

  835:2    warning  Expected linebreaks to be 'CRLF' but found 'LF'  linebreak-style

✖ 835 problems (0 errors, 835 warnings)
  0 errors and 835 warnings potentially fixable with the `--fix` option.

```

Can you fix that?

Then some errors:
```bash
src/extension.ts:59:17 - error TS2339: Property 'registerWebviewViewProvider' does not exist on type 'typeof window'.
59      vscode.window.registerWebviewViewProvider(GitGraphPanelView.viewType, panelProvider)
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/gitGraphPanelView.ts:15:75 - error TS2694: Namespace '"vscode"' has no exported member 'WebviewViewProvider'.
15 export class GitGraphPanelView extends BaseGitGraphView implements vscode.WebviewViewProvider {
                                                                             ~~~~~~~~~~~~~~~~~~~
src/gitGraphPanelView.ts:18:25 - error TS2694: Namespace '"vscode"' has no exported member 'WebviewView'.
18  private _view?: vscode.WebviewView;
                           ~~~~~~~~~~~
src/gitGraphPanelView.ts:85:23 - error TS2694: Namespace '"vscode"' has no exported member 'WebviewView'.
85      webviewView: vscode.WebviewView,
                            ~~~~~~~~~~~
src/gitGraphPanelView.ts:86:19 - error TS2694: Namespace '"vscode"' has no exported member 'WebviewViewResolveContext'.
86      context: vscode.WebviewViewResolveContext,
                        ~~~~~~~~~~~~~~~~~~~~~~~~~

Found 5 errors.
```
I updated my node version and they are gone.
I think wee need to set the minimum version for node. Can you confirm that?

```patch
--- a/package.json
+++ b/package.json
@@ -1633,7 +1633,7 @@
	"devDependencies": {
                "@types/babel__core": "7.1.18",
                "@types/babel__traverse": "7.14.2",
                "@types/jest": "26.0.19",
-               "@types/node": "8.10.62",
+               "@types/node": "^9.4.6",
                "@types/prettier": "<2.6.0",
                "@types/vscode": "1.50.0",
```

LeuciRemi commented:
Oh yes, I was sure I had eliminated those pesky “CRLF” characters 🥲
And yes, updating the Node.js version solves the problem.

hansu commented:
The "Undo Last Commit functionality" does not work well on my system (Ubuntu). It shows the loading popup forever, but when I click "Dismiss" and reload the view, I can see that the action has been performed

![grafik](https://github.com/user-attachments/assets/5dd60c28-6375-44e1-ace8-dd61721b26fa)

Further the wording sounds like it does some magic, maybe I would name it "Undo Last Commit (reset last commit)..." or just "Reset Last Commit..."

hansu commented:
Thanks, will test it later.
But one thing I noticed when building the package - it's size is now 103 MB compared to 3.2 MB before your changes.
Any idea?

hansu commented:
- The "Undo Last Commit functionality" does not work well on my system (Ubuntu). :heavy_check_mark: 

-  'Edit Message' on a commit that is not the HEAD :heavy_check_mark:

LeuciRemi commented:
All right, I'll look at it this weekend.

hansu commented:
1. The selection logic works quite well now, great! But please keep the selection color. Now it's blue and it was grey before.

2. When I select "Drop Selected Commits..." I get a nice list of all commits which shall be dropped.  
Is it possible to get the same list for "Squash Selected Commits..."?
Currently only the first and last are displayed:
![grafik](https://github.com/user-attachments/assets/77bd2ceb-84a5-45b1-be9d-01a41f58832a)

3. For the "Pull Branch ..." with "Force Update" - can you give an example in which case the force option makes sense? Maybe I just use it wrong ? (https://github.com/hansu/vscode-git-graph/pull/62#issuecomment-3022137146)

LeuciRemi commented:
The "Force Update" option is used when you want to grab the remote version of a branch and completely overwrite your local version.

Typical rebase scenario:
You're working on a feature-login branch. Meanwhile, someone rebases this branch onto main and force-pushes the result. Your local history no longer matches the remote because of the rebase.

Solution:
- Stash your current changes (git stash)
- Use "Force Update" to get the rebased version
- Reapply your stash if you want to keep your changes (git stash pop)

It's easier than trying to resolve history conflicts. You just start fresh from the rebased version.

hansu commented:
Really great now! 

But sorry - except that one with the "Force Update" option. What you wrote is exactly what I thought how it should be used. But it doesn't work in my case.

I dug a bit deeper and found that the problem is that you just take the first remote for the `fetchIntoLocalBranch` function.
But in my case I have several remotes and the first one is not the suitable one. So we have to find a way to determine the correct remote for that branch. Not sure if that is even possible.

hansu commented on web/main.ts:
```diff
@@ -1110,6 +1337,19 @@ class GitGraphView {
+						value: this.config.dialogDefaults.fetchIntoLocalBranch.forceFetch,
+						info: 'Force the local branch to be reset to the remote branch (discard local commits).'
+					}], 'Yes, update', (values) => {
+						runAction({ command: 'fetchIntoLocalBranch', repo: this.currentRepo, remote: this.gitRemotes[0], remoteBranch: refName, localBranch: refName, force: <boolean>values[0] }, 'Updating Branch');
```
This statement:

`remote: this.gitRemotes[0]`
 
In my case the array contains:

gitRemotes: [    "bendera",    "origin",    "qrakee",    "upstream"]

but my local branch is only on origin. So it fails.

hansu commented:
Two options:

- We could change the checkbox to use the --rebase option (`git pull --rebase`'). I think this does the same in most cases (if there are no conflicts..).
- The clean solution would be to check before which remote the branch is tracking.
This could be done with this command:
```bash
$ git status -b --porcelain
## hansu-develop...origin/hansu-develop [ahead 2, behind 2]
```
or  
```bash
$ git status -b --porcelain=v2
# branch.oid 8c2a1171a7ac26d6e1655759612a63887a7dd5be
# branch.head hansu-develop
# branch.upstream origin/hansu-develop
```

LeuciRemi commented:
Hi @hansu,

Thanks for the detailed analysis! I've implemented a fix for the "Force Update" issue with multiple remotes.

Solution:
- Added getRemoteForBranch(branchName) function that uses existing gitConfig.branches[branchName].remote data
- Updated Pull Branch dialog to use the correct tracking remote instead of this.gitRemotes[0]
- Added error handling for branches without upstream tracking

Your suggestions (git status -b, git rev-parse @{u}, etc.) would work, but this solution is more efficient since we already have the branch
configuration data loaded. No additional Git calls needed, better performance, same reliability.

hansu commented:
Hi @LeuciRemi,

thanks for your fix! Works fine now :+1:.
Finally time to make a release to get your nice features in!

hansu commented:
@LeuciRemi I noticed a little issue which I think was introduced with this PR.

When I switch to another tab and back to Git Graph again, the selection becomes blue (like in you first approach with multi-selection):
![gg-blue-selection](https://github.com/user-attachments/assets/756db5ea-64f3-4a7b-bffb-e61a69ea305a)

It would be nice if you can investigate this and maybe provide a fix.

Another thing related to the switching I noticed:
The commit selection gets lost when you switch to another tab and back when the CDV is not displayed.
Not sure if this is also related to the above mentioned issue, but it would be nice if you can have a look at this as well.
(I have have not the time for that atm unfortunately)
![gg-selection](https://github.com/user-attachments/assets/0705bbc4-c388-418b-be10-212ecb0be2bc)
